lib/util/vector_template.cpp

```diff
@@ -0,0 +1,1768 @@
+    this->assert_writable();
+    if ((index >= 0) && (index < this->size) && (block_size >= 0)) {
+        if (index + block_size > this->size)
+            block_size = this->size - index;
+
+        index += block_size;
+        this->memmove(index, -block_size);
+    } else
+        util_abort("%s: invalid input \n", __func__);
+}
+
+/**
+   Removes element @index from the vector, shifting all elements to
+   the right of @index one element to the left and shrinking the total
+   vector. The return value is the value which is removed.
+*/
+
+template <typename T> T vector<T>::idel(int index) {
+    T del_value = this->iget(index);
+    this->idel_block(index, 1);
+    return del_value;
+}
+
+/**
+   Removes all occurences of @value from the vector, thereby shrinking
+   the vector. The return value is the number of elements removed from
+   the vector.
+*/
+
+template <typename T> T vector<T>::del_value(T del_value) {
+    int index = 0;
+    int del_count = 0;
+    while (true) {
+        if (index == this->size)
+            break;
+
+        if (this->iget(index) == del_value) {
+            this->idel(index);
+            del_count++;
+        } else
+            index++;
+    }
+    return del_count;
+}
+
+template <typename T> void vector<T>::insert(int index, T value) {
+    if (index >= this->size)
+        this->iset(index, value);
+    else {
+        this->memmove(index, 1);
+        this->iset(index, value);
+    }
+}
+
+template <typename T> void vector<T>::append(T value) {
+    int size = this->size;
+    this->resize(size + 1, 0);
+    this->data[size] = value;
+}
+
+template <typename T> void vector<T>::reset() {
+    this->assert_writable();
+    this->size = 0;
+}
+
+/**
+   The pop function will remove the last element from the vector and
+   return it. If the vector is empty - it will abort.
+*/
+template <typename T> T vector<T>::pop() {
+    this->assert_writable();
+
+    if (this->size > 0) {
+        T value = this->data[this->size - 1];
+        this->size--;
+        return value;
+    } else {
+        util_abort("%s: trying to pop from empty vector \n", __func__);
+        return -1; /* Compiler shut up. */
+    }
+}
+
+template <typename T> void vector<T>::rshift(int shift) {
+    if (shift < 0)
+        this->memmove(-shift, shift);
+    else {
+        this->memmove(0, shift);
+        for (int i = 0; i < shift; i++)
+            this->data[i] = this->default_value;
+    }
+}
+
+template <typename T> void vector<T>::lshift(int shift) {
+    this->rshift(-shift);
+}
+
+template <typename T> void *vector<T>::get_ptr() {
+    this->assert_writable();
+    return this->data;
+}
+
+template <typename T> const void *vector<T>::get_const_ptr() const {
+    return this->data;
+}
+
+/**
+   Observe that there is a principle difference between the get_ptr()
+   functions and alloc_data_copy() when the vector has zero size. The
+   former functions will always return valid (@TYPE@ *) pointer,
```

Comment:
The `(@TYPE@ *)` loooks weird now. Should at least be updated to `T*`. I think it the entire `alloc_data_copy` and its documentation should just be removed. code scars aren't really that useful in my opinion, and we are seriously unlikely to want to add that symbol to the public API at this point.